src/main/scala/basics/DataStructures.scala

```diff
@@ -0,0 +1,86 @@
+package basics
+
+import scala.util.Try
+
+object DataStructures {
+
+  // Exercise. Write a function that checks if all values in a `List` are equal.
+  // Think about what you think your function should return if `list` is empty, and why.
+  def allEqual[T](list: List[T]): Boolean =
+    if (list.isEmpty) true
+    else list.foldLeft((true, list.head)) {
+      case ((previousCheck, previousElement), currentElement) =>
+        (previousCheck && (previousElement == currentElement ), currentElement)
+    }._1
+
+
+  val vegetableWeights = Map(
+    ("pumpkins", 10),
+    ("cucumbers", 20),
+    ("olives", 2),
+  )
+
+  val vegetablePrices = Map(
+    "tomatoes" -> 4,
+    "peppers" -> 5,
+    "olives" -> 17,
+  )
+
+  val vegetableAmounts = Map(
+    "tomatoes" -> 17,
+    "peppers" -> 234,
+    "olives" -> 32,
+    "cucumbers" -> 323,
+  )
+
+  // Exercise. Calculate the total cost of all vegetables, taking vegetable amounts (in units) from
+  // `vegetableAmounts` and prices per unit from `vegetablePrices`. Assume the price is 10 if not available
+  // in `vegetablePrices`.
+  val totalVegetableCost: Int = {
+    vegetableAmounts.toSeq.map {
+      case (vegetable, amount) => amount * vegetablePrices.getOrElse(vegetable, 10)
+    }.sum
+  }
+
+  // Exercise. Given the vegetable weights (per 1 unit of vegetable) in `vegetableWeights` and vegetable
+  // amounts (in units) in `vegetableAmounts`, calculate the total weight per type of vegetable, if known.
+  //
+  // For example, the total weight of "olives" is 2 * 32 == 64.
+  val totalVegetableWeights: Map[String, Int] = { // implement here
+    vegetableWeights.collect {
+      case (vegetable, amount) if vegetableAmounts.contains(vegetable)  =>
+        vegetable -> (vegetableAmounts.getOrElse(vegetable, 0) * amount)
```

Comment:
if you have `getOrElse` why do you need if

Reply:
I need ```if`` for ```collect``` method to be sure that the resulting collection will not have vegetables if there is no information about the amount. I replaced ```getOrElse(vegetable, 0)``` by ```(vegetable)``` , because we know that amount exist.